test: Construct Verifiable Credential object (encoding side)

kdimak opened the pull request:
Here we add examples of VC encoding. VC base model extension is shown by 2 ways:
- struct embedding
- using `ExtraFields` map

Struct embedding does not work properly for now (#847 is created), so it's recommended to use `ExtraFields` map for extensions.

closes #831

Signed-off-by: Dima <dkinoshenko@gmail.com>

kdimak commented on pkg/doc/verifiable/credential.go:
```diff
@@ -674,9 +674,11 @@ func (vc *Credential) JWTClaims(minimizeVC bool) (*JWTCredClaims, error) {
-func (vc *Credential) SubjectID() (string, error) {
+// subjectID gets ID of single subject if present or
+// returns error if there are several subjects or one without ID defined
+func subjectID(subject interface{}) (string, error) {
```
I found the issue with `subjectID()` as it did not support custom structures case. So I had to fix this in this PR to enable the VC encoding example with custom `UniversityCredentialSubject` subject.

haardikk21 commented:
LGTM :+1:

troyronda commented:
Do we have a better name for "Extra Fields" @haardikk21 @kdimak @Windforcer

talwinder50 commented on pkg/doc/verifiable/example_credential_test.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+//nolint:gochecknoglobals
+var privKey = ed25519.PrivateKey{56, 237, 176, 143, 247, 162, 167, 111, 85, 161, 158, 14, 243, 173, 144, 51, 157, 109, 155, 228, 77, 170, 238, 85, 220, 144, 158, 51, 14, 40, 153, 141, 193, 179, 12, 234, 125, 193, 60, 56, 198, 150, 80, 93, 30, 58, 14, 152, 205, 6, 50, 98, 125, 212, 65, 17, 15, 11, 230, 3, 226, 187, 7, 89} //nolint:lll
```
Why we are creating the priv key this way ?

kdimak replied:
If we generate the key, we will get different JWS each time we are running the example. And we are checking the JWS output (2-nd **Outcome**). That's why I made the keys static.

talwinder50 commented on pkg/doc/verifiable/credential.go:
```diff
@@ -674,9 +674,11 @@ func (vc *Credential) JWTClaims(minimizeVC bool) (*JWTCredClaims, error) {
+// subjectID gets ID of single subject if present or
 // returns error if there are several subjects or one without ID defined
-func (vc *Credential) SubjectID() (string, error) {
+// subjectID gets ID of single subject if present or
```
duplication of line , subjectID gets ID of single subject if present

kdimak replied:
Ah, I will fix that! thanks!

kdimak commented:
> Do we have a better name for `ExtraFields` @haardikk21 @kdimak @Windforcer

Does `CustomFields`sound better?

kdimak commented:
I don't know how to reach 85% of patch diff (now it's about 82%)... 
I've changed `subjectID()` in prod code to fix the issue which blocked the successful run of the example, added a test case, but cannot cover all the lines,

haardikk21 commented:
> > Do we have a better name for `ExtraFields` @haardikk21 @kdimak @Windforcer
> 
> Does `CustomFields`sound better?

Yeah `CustomFields` seems more in line with the new structure.

troyronda commented:
when I read the spec, I would understand this `ExtraFields` to be extension claims (and claims also seems to called properties once in the extensions section). Is this correct?

kdimak commented:
> when I read the spec, I would understand this `ExtraFields` to be extension claims (and claims also seems to called properties once in the extensions section). Is this correct?

At [Extensibility section](https://www.w3.org/TR/vc-data-model/#extensibility), they use `new terms` for 2 new fields.  I didn't find any proper term in the spec, that's why we are trying to find the clearest term ourselves. 
For me, "claim" is a whole thing about VC (e.g. "Issuer is a role an entity performs by asserting claims about one or more subjects") or specific JWT term.

kdimak commented:
I changed `ExtraFields` to `CustomFields`. Also found a way how to make the patch diff happy :)

troyronda commented:
Is there a followup test demonstrating signing of the created credential?

haardikk21 commented on pkg/doc/verifiable/common.go:
```diff
@@ -73,7 +73,7 @@ func (tid *TypedID) MarshalJSON() ([]byte, error) {
 
 	alias := (*Alias)(tid)
 
```
We should change the function name to match `CustomFields`

haardikk21 commented on pkg/doc/verifiable/common.go:
```diff
@@ -88,9 +88,9 @@ func (tid *TypedID) UnmarshalJSON(data []byte) error {
+	tid.CustomFields = make(CustomFields)
 
-	err := unmarshalWithExtraFields(data, alias, tid.ExtraFields)
+	err := unmarshalWithExtraFields(data, alias, tid.CustomFields)
```
Same thing for this function.

haardikk21 commented on pkg/doc/verifiable/common.go:
```diff
@@ -73,7 +73,7 @@ func (tid *TypedID) MarshalJSON() ([]byte, error) {
 	alias := (*Alias)(tid)
 
-	data, err := marshalWithExtraFields(alias, tid.ExtraFields)
+	data, err := marshalWithExtraFields(alias, tid.CustomFields)
```
We should change this to match `CustomFields`

kdimak replied:
I was thinking about this renaming and came to the conclusion that `marshalWithExtraFields()` reflects better the purpose of the function - marshall a struct with adding extra fields to the JSON.

haardikk21 replied:
If we're replacing the term "extra fields" with "custom fields" everywhere, I think this should be renamed for consistency reasons. It's also equally valid t say the function Marshalls a struct and adds custom fields to the JSON.

kdimak replied:
Completed the renaming! Now it's more consistent, thanks!

kdimak commented:
> Is there a followup test demonstrating signing of the created credential?

yes, just created #856  covering this demand.
